src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/FixedWindowRateLimiter.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that refreshes allowed permits in a window periodically.
+    /// </summary>
+    public sealed class FixedWindowRateLimiter : ReplenishingRateLimiter
+    {
+        private int _requestCount;
+        private int _queueCount;
+        private long _lastReplenishmentTick;
+        private long? _idleSince;
+        private bool _disposed;
+
+        private readonly Timer? _renewTimer;
+        private readonly FixedWindowRateLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        private object Lock => _queue;
+
+        private static readonly RateLimitLease SuccessfulLease = new FixedWindowLease(true, null);
+        private static readonly RateLimitLease FailedLease = new FixedWindowLease(false, null);
+        private static readonly double TickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
+
+        /// <inheritdoc />
+        public override TimeSpan? IdleDuration => _idleSince is null ? null : new TimeSpan((long)((Stopwatch.GetTimestamp() - _idleSince) * TickFrequency));
+
+        /// <inheritdoc />
+        public override bool IsAutoReplenishing => _options.AutoReplenishment;
+
+        /// <inheritdoc />
+        public override TimeSpan ReplenishmentPeriod => _options.Window;
+
+
+
+        /// <summary>
+        /// Initializes the <see cref="FixedWindowRateLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="FixedWindowRateLimiter"/>.</param>
+        public FixedWindowRateLimiter(FixedWindowRateLimiterOptions options!!)
```

Comment:
The `!!` is actually being removed, see https://github.com/dotnet/runtime/pull/68178
```suggestion
        public FixedWindowRateLimiter(FixedWindowRateLimiterOptions options)
```